modules/imgcodecs/src/loadsave.cpp

```diff
@@ -284,6 +306,50 @@ static void ApplyExifOrientation(const String& filename, Mat& img)
     }
 }
 
+static void ApplyExifOrientation(const String& filename, Mat& img)
+{
+    int orientation = IMAGE_ORIENTATION_TL;
+
+    if (filename.size() > 0)
+    {
+       std::ifstream stream( filename.c_str(), std::ios_base::in | std::ios_base::binary );
+       ExifReader reader( stream );
+       if( reader.parse() )
+       {
+           ExifEntry_t entry = reader.getTag( ORIENTATION );
+           if (entry.tag != INVALID_TAG)
+           {
+               orientation = entry.field_u16; //orientation is unsigned short, so check field_u16
+           }
+       }
+       stream.close();
+    }
+
+    ExifTransform(orientation, img);
+}
+
+static void ApplyExifOrientation(const Mat& buf, Mat& img)
+{
+    int orientation = IMAGE_ORIENTATION_TL;
+
+    if( buf.isContinuous() )
+    {
+       ByteStreamBuffer bsb( reinterpret_cast<char*>(buf.data), buf.total() * buf.elemSize() );
+       std::istream stream( &bsb );
```

Comment:
Looks like the current `ByteStreamBuffer` implementation is buggy.

Please use this instead (with added `#include <sstream>` and removed `ByteStreamBuffer`):
```
std::istringstream stream(std::string(reinterpret_cast<char*>(buf.data), buf.total() * buf.elemSize()), std::ios_base::in);
```

(it does extra `memcpy` operation, but it is not critical for I/O tasks)